### ISSUE TITLE
find_max sign bug

### ISSUE THREAD

**Issue by DavidGEklund** [davisking/dlib#1908]
## Expected Behavior
When using find_max and you end up in a situation where objective function does not change, the objective delta stop strategy should stop the search.

## Current Behavior
when using the dlib::find_max function with dlib::newton_search_strategy and dlib::objective_delta_stop_strategy(1e-7) I end up in a case when the step increment (alpha variable) is zero, but in the stop_strategy.should_continue_search, this is not detected since the f_value sign is toggled with each iteration. The result is an infinite loop. 

## Steps to Reproduce
Provide a non-converging function to find_max with dlib::newton_search_strategy and dlib::objective_delta_stop_strategy where the max != 0.

* **Version**: 
19.18
* **Where did you get dlib**: 
dlib.net
* **Platform**: 
windows 10, 64-bit (not OS dependent)
* **Compiler**: 
msvc 19.16.27034 (not compiler dependent)

**Comment by davisking** [comment]
I'm not sure I see how this can happen. Can you provide a short and complete example that when executed reproduces the problem?

**Comment by DavidGEklund** [comment]
I would love to give you a simple example or just fix the code for you (I normally try to with libraries I use), but I’m unfortunately to pressed for time right now. I solved the problem by using find_min instead of find_max. 

I can however explain in greater detail what happens. In the find_max function stop_strategy.should_continue_search should stop the case I got into (where the alpha variable is 0 and the algorithm is struck), but in this case, the following lines:

if (std::abs(funct_value - _prev_funct_value) < _min_delta)
                    return false;

do not return false because funct_value and prev_funct_value have different signs, and in every iteration of the loop in find_max they both swap signs so that one is always positive and the other negative. It’s a problem only in find_max implementation not present in find_min.

It is an artifact of the sign switching done in the find_max. The alpha is 0 so you’re stuck, and every loop you call: 
f_value *= -1;
switching signs on the function value and in each iteration that is compared to the previous value stored in the stop strategy which has opposite sign.

I hope this is enough to help. I’m sincerely sorry that I don’t have time to just push a fix for you.
